crypto/bls12381.go

```diff
@@ -0,0 +1,166 @@
+		return nil, errors.Wrapf(ErrPrivateKey, "invalid private key length: got %d, want %d", len(b), BLSSecretKeyLength)
+	}
+	if IsZero(b) {
+		return nil, errors.Wrapf(ErrPrivateKey, "private key is zero")
+	}
+	sk := new(blst.SecretKey).Deserialize(b)
+	if sk == nil {
+		return nil, errors.Wrapf(ErrPrivateKey, "invalid private key")
+	}
+	return &BLS12381PrivateKey{p: sk}, nil
+}
+
+// Bytes returns the byte representation of the BLS12381 private key.
+func (k *BLS12381PrivateKey) Bytes() []byte {
+	return k.p.Serialize()
+}
+
+// HexString returns the hexadecimal string representation of the BLS12381 private key.
+func (k *BLS12381PrivateKey) HexString() string {
+	return hex.EncodeToString(k.Bytes())
+}
+
+// Sign signs the given message using the BLS12381 private key.
+func (k *BLS12381PrivateKey) Sign(msg []byte) ([]byte, error) {
+	signature := new(blstSignature).Sign(k.p, msg, dst)
+	return signature.Compress(), nil
+}
+
+// PublicKey returns the public key corresponding to the BLS12381 private key.
+func (k *BLS12381PrivateKey) PublicKey() *BLS12381PublicKey {
+	return &BLS12381PublicKey{
+		p: new(blstPublicKey).From(k.p),
+	}
+}
+
+// Zero clears the BLS12381 private key, effectively zeroizing it.
+func (k *BLS12381PrivateKey) Zero() {
+	k.p.Zeroize()
+}
+
+// BLS12381PublicKeyFromBytes creates a BLS12381 public key from the given byte slice.
+func BLS12381PublicKeyFromBytes(b []byte) (*BLS12381PublicKey, error) {
+	if len(b) != BLSPubkeyLength {
+		return nil, errors.Wrapf(ErrPublicKey, "invalid public key length: got %d, want %d", len(b), BLSPubkeyLength)
+	}
+	pk := new(blstPublicKey).Uncompress(b)
+	if pk == nil {
+		return nil, errors.Wrapf(ErrPublicKey, "invalid public key")
+	}
+	if !pk.KeyValidate() {
+		return nil, errors.Wrapf(ErrPublicKey, "invalid public key, key validation failed")
+	}
+	return &BLS12381PublicKey{p: pk}, nil
+}
+
+// Bytes returns the byte representation of the BLS12381 public key.
+func (k *BLS12381PublicKey) Bytes() []byte {
+	return k.p.Compress()
+}
+
+// HexString returns the hexadecimal string representation of the BLS12381 public key.
+func (k *BLS12381PublicKey) HexString() string {
+	return hex.EncodeToString(k.Bytes())
+}
+
+// Verify verifies the given signature against the message using the BLS12381 public key.
+func (k *BLS12381PublicKey) Verify(msg []byte, sig []byte) bool {
+	signature := new(blstSignature).Uncompress(sig)
+	if signature == nil {
+		return false
+	}
+	return signature.Verify(true, k.p, false, msg, dst)
+}
+
+// BLSAggregateSignature aggregates multiple BLS signatures into a single signature.
+func BLSAggregateSignature(sigs [][]byte) ([]byte, error) {
+	signature, err := blsAggregateSignature(sigs)
+	if err != nil {
+		return nil, err
+	}
+	return signature.Compress(), nil
+}
+
+// BLSAggregateVerify verifies multiple BLS signatures against the message using the provided public keys.
+func BLSAggregateVerify(pubKeys [][]byte, sigs [][]byte, msg []byte) (bool, error) {
```

Comment:
the second parameter `sigs` should be the aggregated signature `aggrSign []byte`